src/factory/mastercopyDeployer.ts

```diff
@@ -7,51 +7,115 @@ import { getSingletonFactory } from "./singletonFactory";
  *
  * To get the same address on any chain.
  * @param hre hardhat runtime environment
- * @param mastercopyContractFactory
- * @param args
- * @returns The address of the deployed module mastercopy
+ * @param mastercopyContractFactory mastercopy to deploy
+ * @param args the arguments to pass to the mastercopy's constructor
+ * @returns The address of the deployed module mastercopy or undefined if it was already deployed
  */
 export const deployMastercopy = async (
   hre: HardhatRuntimeEnvironment,
   mastercopyContractFactory: ContractFactory,
   args: Array<any>,
   salt: string
-) => {
+): Promise<string | undefined> => {
   const deploymentTx = mastercopyContractFactory.getDeployTransaction(...args);
+
   if (deploymentTx.data) {
-    await deployMastercopyWithInitData(hre, deploymentTx.data, salt);
+    return await deployMastercopyWithInitData(hre, deploymentTx.data, salt);
+  }
+  throw new Error("No deployment data found");
+};
+
+/**
+ * Compute a module's mastercopy address. Where it is or will be deployed.
+ *
+ * @param hre hardhat runtime environment
+ * @param mastercopyContractFactory mastercopy to get address for
+ * @param args the arguments passed to the mastercopy's constructor
+ * @returns {
+ *  address: string; // the address where the module mastercopy will be deployed or was already deployed
+ *  isDeployed: boolean; // true if the module mastercopy was already deployed on this chain
+ * }
+ */
+export const computeTargetAddress = async (
+  hre: HardhatRuntimeEnvironment,
+  mastercopyContractFactory: ContractFactory,
+  args: Array<any>,
+  salt: string
+): Promise<{ address: string; isDeployed: boolean }> => {
+  const deploymentTx = mastercopyContractFactory.getDeployTransaction(...args);
+  const singletonFactory = await getSingletonFactory(hre);
+
+  if (!deploymentTx.data) {
+    throw new Error("No deployment data found");
   }
+
+  const initCodeHash = hre.ethers.utils.solidityKeccak256(
+    ["bytes"],
+    [deploymentTx.data]
+  );
+  const computedAddress = hre.ethers.utils.getCreate2Address(
+    singletonFactory.address,
+    salt,
+    initCodeHash
+  );
+
+  const targetAddress = utils.getAddress(
+    (await singletonFactory.callStatic.deploy(
+      deploymentTx.data,
+      salt
+    )) as string
+  );
+
+  // Sanity check
+  if (
+    computedAddress !== targetAddress &&
+    targetAddress !== "0x0000000000000000000000000000000000000000"
+  ) {
```

Comment:
This right here should be an assertion and not an error. Reason: the assertion is introduced solely to guard against a potential future programming error introduced in the way we code the address compute.

Essentially there's no reason why the would differ, other than a programming error. I would write this as:

assert(targetAddress === AddressZero || targetAddress === computedAddress)

Reply:
I totally agree, I actually implemented it with an assert first but then the editor pulled in `assert` from `console`, and that seems to only be logging an error (https://developer.mozilla.org/en-US/docs/Web/API/console/assert). It was unclear if the program would halt.
For the case of NodeJs assert, it looks like it is intended for internal use (in NodeJs) and should not be used in production.

Then I asked, ChatGPT:

> ... It's important to note that the assert module is intended for internal testing and debugging, and is not meant to be used for handling expected errors in production code. For example, you might want to use a try-catch block to handle the error ...

Perhaps it's still okay to use NodeJs assert in this script? 🤔

Reply:
ahah! ChatGPT the guru.

Well I certainly use the assert from nodejs all the time! I'm a big assert fan.